flexmeasures/data/models/reporting/__init__.py

```diff
@@ -0,0 +1,170 @@
+                event_ends_before=event_ends_before,
+                **tb_query,
+            )
+
+            # adding sources
+            for source in bdf.sources.unique():
+                self.data[f"source_{source.id}"] = source
+
+            # saving bdf
+            self.data[
+                f"sensor_{sensor.id}"
+            ] = bdf  # TODO: Add alias to reference dataframes easily. e.g: dict(sensor = 1, alias="power"),
+
+    def compute(self, *args, **kwargs) -> Optional[pd.DataFrame]:
+        """This method triggers the creation of a new report. This method allows to update the fields
+        in reporter_config_raw passing them as keyword arguments or the whole `reporter_config_raw` by
+        passing it in the kwarg `reporter_config_raw`.
+
+            Overall, this method follows these steps:
+                1) Updating the reporter_config with the kwargs of the method compute.
+                2) Triggers config deserialization.
+                3) Fetches the data of the sensors described by the field `tb_query_config`.
+                4) If the output is BeliefsDataFrame, it simplifies it into a DataFrame
+
+        """
+        # if report_config in kwargs
+        if "reporter_config_raw" in kwargs:
+            self.reporter_config_raw.update(kwargs.get("reporter_config_raw"))
+        else:  # check for arguments in kwarg that could be potential fields of reporter config
+            for key, value in kwargs.items():
+                if key in self.reporter_config_raw:
+                    self.reporter_config_raw[key] = value
+
+        # deserialize configuration
+        self.deserialize_config()
+
+        # fetch data
+        self.fetch_data()
+
+        # Result
+        result = self._compute()
+
+        if isinstance(result, tb.BeliefsDataFrame):
+            result = simplify_index(result)
+
+        return result
+
+    def _compute(self) -> Optional[pd.DataFrame]:
+        """
+        Overwrite with the actual computation of your report.
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    def get_data_source_info(cls: type) -> dict:
+        """
+        Create and return the data source info, from which a data source lookup/creation is possible.
+        See for instance get_data_source_for_job().
+        """
+        source_info = dict(
+            model=cls.__name__, version="1", name="Unknown author"
+        )  # default
+
+        if hasattr(cls, "__version__"):
+            source_info["version"] = str(cls.__version__)
+        else:
+            current_app.logger.warning(
+                f"Scheduler {cls.__name__} loaded, but has no __version__ attribute."
+            )
+        if hasattr(cls, "__author__"):
+            source_info["name"] = str(cls.__author__)
+        else:
+            current_app.logger.warning(
+                f"Scheduler {cls.__name__} has no __author__ attribute."
+            )
+        return source_info
```

Comment:
I'd still like to see this refactored so that both Scheduler and Reporter share this code. Either using a mixin (`DataGenerator`/`DataSourcerer` ;) ?) or a shared utility function (`_get_data_source_info(cls)`?).